docs/architecture.md

```diff
@@ -0,0 +1,222 @@
+
+## High Level design
+
+The high level flow for the w3filecoin Pipeline is:
+
+- **Event** is triggered once a CAR file is written into a bucket with its metadata {`link`, `size`, `bucketName`, `bucketEndpoint`}. This event is added to a `content_validator_queue`.
+- **Content Validator process** validates CARs and writes references to a `content` table.
+- On its own schedule, **Piece maker process** can pull queued content from a `content_queue`, derive piece info for them and write records to a `piece` & `inclusion` tables.
+  - a `inclusion` table enables same piece to be included in another aggregate if a deal fails.
+- **Agregagtor process** reads from a `cargo_queue` (backed by `inclusion` table), attempts to create an aggregate and if successful it writes to an `aggregate` table.
+- **Submission process** reads from the `aggregate_queue`, submits aggregates to the agency (spade proxy) and writes deal record with status "PENDING".
+- TBD deal flow
+
+![Pipeline processes](./processes.png)
+
+The w3filecoin pipeline is modeled into 4 different SST Stacks that will have their infrastructure provisioned in AWS via AWS CloudFormation. These are:
+
+- API Stack
+- Consumer Stack
+- DB Stack
+- Processor Stack
+
+![Architecture](./architecture.png)
+
+## API Stack
+
+TBC
+API Gateway to expose:
+- Report API for failed aggregates to land into Storage Providers
+- Get to know state of deals
+- ...
+
+## Consumer Stack
+
+w3filecoin relies on events from Buckets once CAR files are written into them. It is designed to enable multiple sources of CAR files to be easily integrated into the pipeline via its consumer stack.
+
+As an example, w3filecoin is wired up with `w3infra` as a source of CAR files to get into Filecoin deals. `w3infra` will emit events once CAR files are written into desired buckets. This events should include necessary information of the location of the CAR file to enable w3filecoin to let Storage Providers know where they can fetch the CAR files from.
+
+Further down the line, consumer stack can be wired with the UCAN Log Stream and use receipts as the trigger.
+
+## DB Stack
+
+The DB stack relies on Amazon RDS Database to keep the necessary state for the w3filecoin pipeline. Its data model was designed with the aim of being the data structure for each of the processors running within the pipeline, while also enabling the tracking of state of each item in the pipeline and to get a mapping between content CIDs and piece CIDs.
+
+### Schema
+
+```sql
+-- Table describes queue of verified CARs to be stored in filecoin
+-- CAR is considered in queue when there is no piece referencing it
+CREATE TABLE content
+(
+  -- CAR CID
+  link TEXT PRIMARY KEY,
+  -- CAR Size
+  size number NOT NULL,
+  -- Source where the content can be fetched from
+  source JSONB NOT NULL,
```

Comment:
I'm not sure I understand what JSON document look like here, mind adding a comment so it's clear. We also may have to consider what to do if malformed record is encountered (which seems more likely with arbitrary JSONs over time) so that whole processing pipeline does not get stack if one bad record makes it into the table.

Reply:
I added. And yes, whoever writes the content should validate this (consumer stack)